src/main/java/org/openstreetmap/atlas/geography/boundary/DynamicGridIndexBuilder.java

```
@@ -1,40 +1,223 @@
+                    // at least the boundary itself. If query results isSameCountry returns true,
+                    // the box must ONLY intersect with the country the boundary belongs, in this
+                    // case we treat it same as inside cell. This is an end point, we add the box to
+                    // the spatial index.
+                    synchronized (this.index)
+                    {
+                        this.index.insert(box, polygon);
+                    }
+                    return;
+                }
+
+                // None of above conditions are met, we need to split the box into sub-boxes
+                if (width > height)
+                {
+                    this.queue.add(new GridWorkItem(minX, minY, minX + width / 2, maxY, polygon));
+                    this.queue.add(new GridWorkItem(minX + width / 2, minY, maxX, maxY, polygon));
+                }
+                else
+                {
+                    this.queue.add(new GridWorkItem(minX, minY, maxX, minY + height / 2, polygon));
+                    this.queue.add(new GridWorkItem(minX, minY + height / 2, maxX, maxY, polygon));
+                }
+            }
+            catch (final TopologyException e)
+            {
+                // Catch mostly JTS Topology exceptions
+                final String countryCode = CountryBoundaryMap.getGeometryProperty(polygon,
+                        ISOCountryTag.KEY);
+                logger.error("Unable to build tree under box {} for country code {}.", box,
+                        countryCode, e);
+                // Save the current box and do not iterate down.
+                synchronized (this.index)
+                {
+                    this.index.insert(box, polygon);
+                }
+            }
+            catch (final Exception e)
+            {
+                logger.error("Processor failed.", e);
```

What's the impact of failure here? Corrupted grid index?

---

Yes, possibly corrupted index. I'll let it throw here.